pkg/server/api_buildimage.go

```diff
@@ -126,21 +143,46 @@ func (i *InventoryServer) apiBuildImagePostStart(resp http.ResponseWriter, req *
 				}
 				return nil
 			}
+			logrus.Debugf("build-image: build job already running with token %s and seed image %s.", sanitizedJob.Token, reg.buildImageURL)
 			errMsg := fmt.Errorf("a build image task is already running for the MachineRegistration")
 			http.Error(resp, errMsg.Error(), http.StatusBadRequest)
 			return errMsg
 		}
 	}
 
+	// Pod lifecycle: currently we allow just one build task at a time that stays around till we don't start another one.
+	// TODO: extend to allow multiple build Pods
+	// TODO: ensure proper clean-up when the operator is removed or upgraded
+	regList := i.registrationCache.getRegistrationKeys()
+	for _, token := range regList {
+		if status, _ := i.registrationCache.getBuildImageStatus(token); status == jobStatusStarted || status == jobStatusInit {
+			errMsg := fmt.Errorf("a build image task is already running for MachineRegistration with token %s", token)
+			http.Error(resp, errMsg.Error(), http.StatusServiceUnavailable)
+			return errMsg
+		} else if status == jobStatusCompleted {
+			logrus.Debugf("build-image: delete completed job with token %s.", token)
+			i.deleteBuildImagePodService(buildImgResName, buildImgResNamespace)
+			_ = i.registrationCache.setBuildImageStatus(token, jobStatusNotStarted)
+			break
+		}
+	}
+
+	// In the unlikely case we have an untracked build Pod (old operator which got upgraded?) let's enforce clean-up
+	pod := &corev1.Pod{}
+	if err := i.Get(i, client.ObjectKey{Name: buildImgResName, Namespace: buildImgResNamespace}, pod); err == nil {
+		logrus.Warnf("build-image: found orphan build pod, enforce clean-up.")
+		i.deleteBuildImagePodService(buildImgResName, buildImgResNamespace)
+	}
+
 	i.registrationCache.setRegistrationData(sanitizedJob.Token, registrationData{
 		buildImageURL:    sanitizedJob.URL,
-		buildImageStatus: jobStatusStarted,
+		buildImageStatus: jobStatusInit,
 		downloadURL:      "",
 	})
 
-	logrus.Infof("New build-image job queued: seed image:'%s', reg token:'%s'", sanitizedJob.URL, sanitizedJob.Token)
+	logrus.Infof("build-image: new job queued (seed image:'%s', reg token:'%s')", sanitizedJob.URL, sanitizedJob.Token)
 	// start the actual build job here
-	go i.draftDoBuildImage(sanitizedJob)
+	go i.doBuildImage(sanitizedJob)
```

Comment:
In a separate thread?  Is there any sync control? I am wondering if should we have a mediaBuild CRD and move job build management into the controller logic and just handle crd creation and tracking from the HTTP end point. So we could somehow also handle timeout for serving the built artifacts and stuff like that.

Reply:
We need a separate thread as we needed a control loop over the build-image Pod. Without a go routine there, it would likely impact the main Elemental Operator functionalities.

Reply:
I completely agree with what you wrote: there are many hints that we need to deal with this with a controller (and so a new CRD for the build-image task).